force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,201 @@
+                DonationHistoryController.DonationHistory softCredit = new DonationHistoryController.DonationHistory();
+                softCredit.id = ocr.Id;
+                softCredit.name = ocr.Opportunity.Name;
+                softCredit.isPrimary = ocr.isPrimary;
+                softCredit.closeDate = ocr.Opportunity.CloseDate;
+                softCredit.role = ocr.Role;
+                softCredit.amount = ocr.Opportunity.Amount;
+                softCredit.creditType = SOFT_CREDIT;
+                softCredit.paymentMethod = 'Cash';
+                softCreditList.add(softCredit);
+            }
+        }
+        return softCreditList;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Filter for ids already classified as Soft Credit.
+    */
+    private static List<Id> filterVerifiedIds(List<DonationHistoryController.DonationHistory> softCredit){
+        List<Id> donationsId = new List<Id>();
+        for(DonationHistoryController.DonationHistory dh: softCredit){
+            donationsId.add(dh.id);
+        }
+        return donationsId;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify partial soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyPartialSoftCreditCondition(List<Id> donationsId, List<OpportunityContactRole> donations){
+        //search on PartialSoftCredit object for ids filtered before usin IN()
+        //only ids with not type of credit defined before
+        List<Id> partialSoftCreditIds = new List<Id>();
+        List<Partial_Soft_Credit__c> partialSoftCredit = [SELECT id, Amount__c, Contact_Role_ID__c FROM Partial_Soft_Credit__c WHERE Contact_Role_ID__c IN :donationsId];
+        for(Partial_Soft_Credit__c ps: partialSoftCredit){
+            partialSoftCreditIds.add(ps.Contact_Role_ID__c);
+        }
+        List<DonationHistoryController.DonationHistory> partialSoftCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (partialSoftCreditIds.contains(ocr.id)) {
+                DonationHistoryController.DonationHistory newPartialSoftCredit = new DonationHistoryController.DonationHistory();
+                newPartialSoftCredit.id = ocr.Id;
+                newPartialSoftCredit.name = ocr.Opportunity.Name;
+                newPartialSoftCredit.isPrimary = ocr.isPrimary;
+                newPartialSoftCredit.closeDate = ocr.Opportunity.CloseDate;
+                newPartialSoftCredit.role = ocr.Role;
+                newPartialSoftCredit.creditType = PARTIAL_SOFT_CREDIT;
+                newPartialSoftCredit.paymentMethod = 'Cash';
+                //TODO: change for Amount__c get on partialSoftCredit query
+                newPartialSoftCredit.amount = ocr.Opportunity.Amount;
+                partialSoftCreditList.add(newPartialSoftCredit);
```

Comment:
Same duplicate logic as noted previously.

Reply:
Constructor was added for the class but the logic that validates the types of credits is different for each case, so it was not modified.